### ISSUE TITLE
[Question] Connecting to Chat

### ISSUE THREAD

**Issue by xlxAciDxlx** [SaviorXTanren/mixer-client-csharp#18]
Probably a stupid question, but how would one go about connecting to chat? I was able to get my bot to connect to Mixer (via Oauth), but I'm unsure of how to connect to chat and parse messages (or at least catch them).

I've tried looking through the unit tests to get an idea, but that just confused me more. I'm sure I'm missing something completely simple to find, but any help would be appreciated. Thanks!

**Comment by xlxAciDxlx** [comment]
Nevermind, I'm an idiot.. Completely missed the "Samples" directory.. Sorry about that!

**Comment by SaviorXTanren** [comment]
Not a problem :) I had to rework the directories recently to get Travis-CI builds working, so all of the samples used to sit at the root directory, but now they are all under the Samples folder. Let me know if you have any other questions :)

**Comment by xlxAciDxlx** [comment]
So, this might be a stupid question, but is there any way to prevent the OAuth localhost page from opening every single time? Obviously, once my bot is done it'll only be opened once and the bot (hopefully :P) won't be restarted constantly, but during development it can be a tad annoying..

**Comment by SaviorXTanren** [comment]
So OAuth page is going to be opened every time you try to get a new OAuth token, that's just the way OAuth works. However, if you have cached OAuth token from the last time, you can re-establish an OAuth connection with that token.

If you look in the Unit Tests, there's a class called MixerConnectionUnitTests. You'll want to look at the ConnectWithOldOAuthToken() method for a quick example.

Basically, when you have created & established a MixerConnection object, you can call GetOAuthTokenCopy() to get a OAuthTokenModel object which you can serialize out somewhere when you're closing your application. When you relaunch your application, deserialize that OAuthTokenModel and pass it in to MixerConnection.ConnectViaOAuthToken(). This will refresh the OAuth and attempt to re-establish a connection. If we can't do it, we'll pop an OAuth window like normally.

**Comment by xlxAciDxlx** [comment]
I knew of the methods, I just couldn't figure out the chain of events to get it working. Got it working now though, thanks! :D

**Comment by xlxAciDxlx** [comment]
Sorry to bug you again ;) But would it be possible to get the channel information included in events that don't currently have it? I'm running a bot that connects to multiple channels, and I don't currently have a way to send a message to only that specific channel that a message was received from (that I'm aware of anyway).

**Comment by SaviorXTanren** [comment]
Hey, I'd ask you to create a separate issue in the future rather than re-using an unrelated, old one. Spent about 5 minutes trying to find which issue you had replied to and helps keep requests more silo-ed.

I'm not sure I follow your request exactly. Are you talking about Constellation events or Chat events? Below is a suggestion based off of what I can gather from your request, but I may be missing something.

If your bot is connecting to multiple channels, you should try creating a class that encapsulates the Channel you're connected to and contains the ChatClient for that channel. In that class, hook up listeners to the events on the ChatClient and then use that same ChatClient when events are received to send your messages.

**Comment by xlxAciDxlx** [comment]
Sorry about that, figured it'd be better to contain my questions on one topic rather than spreading them out over a few ones. But I'll do that next time.

I'm referring to Chat events, I haven't messed with Constellation events yet. Right now, my bot has a single connection and then loops through a string of usernames (essentially chats to connect to) and creates/stores a ChatClient for each. My issue is, when an event is fired there's no easy way to know which channel it came from (mainly for logging/debugging purposes).

**Comment by xlxAciDxlx** [comment]
Yeah, if you could do that I'd _very_ much appreciate it as I'm not quite sure I'm getting what you mean.

**Comment by SaviorXTanren** [comment]
Here you go: [MixerChatConnectionTest.zip](https://github.com/SaviorXTanren/mixer-client-csharp/files/1709800/MixerChatConnectionTest.zip)

ChannelChatClient is the encapsulation class I'm talking about, it keeps track of the channel it's connected to and the ChatClient for that class. Since each class instance is only listening for events from the ChatClient that it has, it'll only receive events for that Channel's chat.

**Comment by xlxAciDxlx** [comment]
Thanks! I guess I was trying to over-simply things by just having a `Dictionary<String, ChatClient>` rather than having a class wrap with all the data I needed.
